src/libstd/collections/hash/table.rs

```diff
@@ -205,41 +193,56 @@ impl<K, V, M> EmptyBucket<K, V, M> {
     pub fn into_table(self) -> M {
         self.table
     }
-}
-
-impl<K, V, M> Bucket<K, V, M> {
-    /// Move out the reference to the table.
-    pub fn into_table(self) -> M {
-        self.table
-    }
     /// Get the raw index.
     pub fn index(&self) -> usize {
         self.idx
     }
+    /// Turn the bucket into an iterator over full buckets.
+    pub fn into_iter(self) -> RawFullBuckets<K, V, M> {
+        RawFullBuckets {
+            raw: self.raw,
+            hashes_end: unsafe {
+                self.raw.hash.offset(self.capacity as isize - self.index() as isize)
+            },
+            table: self.table,
+        }
+    }
 }
 
-impl<K, V, M: Deref<Target=RawTable<K, V>>> Bucket<K, V, M> {
-    pub fn new(table: M, hash: SafeHash) -> Bucket<K, V, M> {
-        Bucket::at_index(table, hash.inspect() as usize)
+impl<K, V, M> Bucket<K, V, M> where M: Borrow<RawTable<K, V>> {
+    pub fn new(table: M, hash: SafeHash) -> Option<Bucket<K, V, M>> {
+        Bucket::at_index(table, *hash as usize).ok()
     }
 
-    pub fn at_index(table: M, ib_index: usize) -> Bucket<K, V, M> {
-        let ib_index = ib_index & (table.capacity() - 1);
-        Bucket {
-            raw: unsafe {
-               table.first_bucket_raw().offset(ib_index as isize)
-            },
-            idx: ib_index,
-            table: table
+    pub fn at_index(table: M, ib_index: uint)
+                   -> Result<Bucket<K, V, M>, Bucket<K, V, M, bucket::TableIsEmpty>>
+    {
+        let capacity = table.borrow().capacity();
+        let idx = ib_index & (capacity - 1);
+        let bucket: Bucket<K, V, M> = Bucket {
+            raw: unsafe { table.borrow().first_bucket_raw().offset(idx as isize) },
```

Comment:
Is it undefined behaviour to call `.offset` on a null pointer? If `capacity == 0`, this is what will happen.

Note: `.offest` translates to an `inbounds` GEP instruction in LLVM.